File: bindings/dotnet/src/Aster/Interceptors/CompressionInterceptor.cs
namespace Aster.Interceptors;

/// <summary>
/// Per-call compression override interceptor.
/// Spec reference: S9.2g
///
/// Injects compression settings into the call metadata so downstream
/// transport code can honour them. Actual compression is performed by
/// the codec/framing layer.
/// </summary>
public class CompressionInterceptor : IInterceptor
{
    /// <summary>Default compression threshold in bytes.</summary>
    public const int DefaultCompressionThreshold = 4096;

    /// <summary>Payload size (bytes) above which compression is applied.</summary>
    public int Threshold { get; }

    /// <summary>Master switch. When false, compression is suppressed.</summary>
    public bool Enabled { get; }

    /// <param name="threshold">
    /// Payload size (bytes) above which compression is applied.
    /// Set to -1 to disable compression regardless of payload size.
    /// Defaults to 4096.
    /// </param>
    /// <param name="enabled">Master switch. When false, compression is suppressed.</param>
    public CompressionInterceptor(int threshold = DefaultCompressionThreshold, bool enabled = true)
    {
        Threshold = threshold;
        Enabled = enabled;
    }

    public object OnRequest(CallContext ctx, object request)
    {
        int effectiveThreshold = Enabled ? Threshold : -1;
        ctx.Metadata["_aster_compress_threshold"] = effectiveThreshold.ToString();
        ctx.Metadata["_aster_compress_enabled"] = Enabled.ToString().ToLowerInvariant();
        return request;
    }
}

File: bindings/java/src/main/java/com/aster/interceptors/CompressionInterceptor.java
package com.aster.interceptors;

/**
 * Per-call compression override interceptor.
 *
 * <p>Injects {@code _aster_compress_threshold} and {@code _aster_compress_enabled} into the call
 * context metadata so downstream transport code can honour per-call compression settings.
 */
public final class CompressionInterceptor implements Interceptor {

  /** Default compression threshold in bytes. */
  public static final int DEFAULT_THRESHOLD = 4096;

  private final int threshold;
  private final boolean enabled;

  /** Creates a compression interceptor with default threshold (4096) and enabled. */
  public CompressionInterceptor() {
    this(DEFAULT_THRESHOLD, true);
  }

  /**
   * Creates a compression interceptor.
   *
   * @param threshold payload size in bytes above which compression is applied. Set to -1 to disable
   *     regardless of payload size.
   * @param enabled master switch; when false, compression is suppressed for every call
   */
  public CompressionInterceptor(int threshold, boolean enabled) {
    this.threshold = threshold;
    this.enabled = enabled;
  }

  @Override
  public Object onRequest(CallContext ctx, Object request) {
    int effectiveThreshold = enabled ? threshold : -1;
    ctx.metadata().put("_aster_compress_threshold", String.valueOf(effectiveThreshold));
    ctx.metadata().put("_aster_compress_enabled", String.valueOf(enabled));
    return request;
  }
}
